src/mono/mono/mini/interp/interp.c

```diff
@@ -7553,13 +7555,66 @@ invalidate_transform (gpointer imethod_)
 	imethod->transformed = FALSE;
 }
 
+static void
+copy_imethod_for_frame (MonoDomain *domain, InterpFrame *frame)
+{
+	InterpMethod *copy = (InterpMethod *) mono_domain_alloc0 (domain, sizeof (InterpMethod));
+	memcpy (copy, frame->imethod, sizeof (InterpMethod));
+	copy->next_jit_code_hash = NULL; /* we don't want that in our copy */
+	frame->imethod = copy;
+	/* Note: The copy will be around until the domain is unloading. Ideally we
+	 * would reclaim its memory when the corresponding InterpFrame is popped.
+	 */
+}
+
+static void
+interp_metadata_update_init (MonoError *error)
+{
+	if ((mono_interp_opt & INTERP_OPT_INLINE) != 0)
+		mono_error_set_execution_engine (error, "Interpreter inlining must be turned off for metadata updates");
+}
+
 static void
 interp_invalidate_transformed (MonoDomain *domain)
 {
+	gboolean need_stw_restart = FALSE;
+#ifdef ENABLE_METADATA_UPDATE
+	need_stw_restart = TRUE;
+
+	mono_gc_stop_world ();
+	/* (1) make a copy of imethod for every interpframe that is on the stack,
+	 * so we do not invalidate currently running methods */
+
+	FOREACH_THREAD_EXCLUDE (info, MONO_THREAD_INFO_FLAGS_NO_GC) {
+		if (!info)
+			continue;
+
+		MonoLMF *lmf = info->jit_data->lmf;
```

Comment:
This fails to detect the currently executing methods (methods that are executing since the last interp entry)

Reply:
@BrzVlad how do I get the ones at the top of the stack?

Reply:
I think the easiest for now would be to save the current frame in the `ThreadContext*` when we have a suspend request while in `MINT_SAFEPOINT`.

Reply:
@BrzVlad done, take another look, please.